python/ray/air/preprocessor.py

```diff
@@ -158,25 +172,101 @@ def _check_is_fitted(self) -> bool:
         fitted_vars = [v for v in vars(self) if v.endswith("_")]
         return bool(fitted_vars)
 
+    @DeveloperAPI
     def _fit(self, dataset: Dataset) -> "Preprocessor":
         """Sub-classes should override this instead of fit()."""
         raise NotImplementedError()
 
+    def _determine_transform_to_use(self, data_format: str) -> str:
+        """Determine which transform to use based on data format and implementation.
+
+        * If only _transform_arrow is implemented, will convert the data to arrow.
+        * If only _transform_pandas is implemented, will convert the data to pandas.
+        If both are implemented, will pick the method corresponding to the format
+        for best performance.
+        * Implementation is defined as overriding the method in a sub-class.
+        """
+
+        assert data_format in ("pandas", "arrow")
+        has_transform_arrow = (
+            self.__class__._transform_arrow != Preprocessor._transform_arrow
+        )
+        has_transform_pandas = (
+            self.__class__._transform_pandas != Preprocessor._transform_pandas
+        )
+
+        if has_transform_arrow and has_transform_pandas:
+            # Both transforms available, so we delegate based on the dataset format to
+            # ensure minimal data conversions.
+            if data_format == "pandas":
+                transform_type = "pandas"
+            elif data_format == "arrow":
+                transform_type = "arrow"
+        elif has_transform_pandas:
+            transform_type = "pandas"
+        elif has_transform_arrow:
+            transform_type = "arrow"
+        else:
+            raise NotImplementedError(
+                "Neither `_transform_arrow` nor `_transform_pandas` are implemented."
+            )
+        return transform_type
+
     def _transform(self, dataset: Dataset) -> Dataset:
         # TODO(matt): Expose `batch_size` or similar configurability.
         # The default may be too small for some datasets and too large for others.
-        return dataset.map_batches(self._transform_pandas, batch_format="pandas")
+
+        dataset_format = dataset._dataset_format()
+        if dataset_format not in ("pandas", "arrow"):
+            raise ValueError(
+                f"Unsupported Dataset format: '{dataset_format}'. Only 'pandas' and "
+                "'arrow' Dataset formats are supported."
+            )
```

Comment:
Shouldn't `simple` be able to be converted to either `pandas` or `arrow`?

Reply:
Talked about it with Clark and it's going to be eventually handled properly on the dataset side.

Reply:
We could do a best-effort conversion, but `simple` is often used as an escape hatch for types that are not representable in Arrow/Pandas.

Reply:
Simple can always be represented in "pandas" I believe.